crate/utils/src/crypto/hybrid_encryption/ecies.rs

```diff
@@ -0,0 +1,204 @@
+use num_bigint_dig::algorithms::idiv_ceil;
+#[cfg(feature = "fips")]
+use openssl::nid::Nid;
+use openssl::{
+    bn::BigNumContext,
+    ec::{EcGroupRef, EcKey, EcPoint, EcPointRef, PointConversionForm},
+    hash::{Hasher, MessageDigest},
+    pkey::{PKey, Private, Public},
+    symm::{decrypt_aead, encrypt_aead, Cipher},
+};
+
+#[cfg(feature = "fips")]
+use crate::kmip_utils_bail;
+use crate::{
+    crypto::symmetric::{AES_256_GCM_IV_LENGTH, AES_256_GCM_KEY_LENGTH, AES_256_GCM_MAC_LENGTH},
+    error::KmipUtilsError,
+};
+
+/// Derive initialization vector from recipient public key `Q` and ephemeral
+/// public key `R` using SHAKE128-128.
+#[allow(non_snake_case)]
+fn ecies_get_iv(
+    Q: &EcPointRef,
+    R: &EcPointRef,
+    curve: &EcGroupRef,
+) -> Result<Vec<u8>, KmipUtilsError> {
+    let mut ctx = BigNumContext::new_secure()?;
+    let Q_vec = Q.to_bytes(curve, PointConversionForm::COMPRESSED, &mut ctx)?;
+    let R_vec = R.to_bytes(curve, PointConversionForm::COMPRESSED, &mut ctx)?;
+
+    let mut iv = vec![0; AES_256_GCM_IV_LENGTH];
+
+    let mut hasher = Hasher::new(MessageDigest::shake_128())?;
+    hasher.update(&Q_vec)?;
+    hasher.update(&R_vec)?;
+    hasher.finish_xof(&mut iv)?;
+
+    Ok(iv)
+}
+
+/// Derive S into the symmetric secret key using SHAKE128-256.
+#[allow(non_snake_case)]
+fn ecies_get_key(S: &EcPointRef, curve: &EcGroupRef) -> Result<Vec<u8>, KmipUtilsError> {
+    let mut ctx = BigNumContext::new_secure()?;
+    let S_vec = S.to_bytes(curve, PointConversionForm::COMPRESSED, &mut ctx)?;
+
+    let mut key = vec![0; AES_256_GCM_KEY_LENGTH];
+
+    let mut hasher = Hasher::new(MessageDigest::shake_128())?;
+    hasher.update(&S_vec)?;
+    hasher.finish_xof(&mut key)?;
+
+    Ok(key)
+}
+
+/// Encrypt `plaintext` data using `pubkey` public key following the ECIES
+/// principle.
```

Comment:
I though the KMS used the `pk` abbreviation. My bad since it does not seem to be the case. My comment still holds for the use of principle though. ECIES is not a principle, it is an encryption scheme: you are either following it or not. Principle is too vague.

Reply:
e20c6e52348576621a9499c906cff9e1f83eaa62